v2/yarpcrouter/router.go

```diff
@@ -58,19 +59,78 @@ type MapRouter struct {
 func NewMapRouter(defaultService string) MapRouter {
 	return MapRouter{
 		defaultService:            defaultService,
-		serviceProcedureEncodings: make(map[serviceProcedureEncoding]yarpc.Procedure),
+		serviceProcedureEncodings: make(map[serviceProcedureEncoding]yarpc.TransportProcedure),
 		serviceNames:              map[string]struct{}{defaultService: {}},
 	}
 }
 
+// MapToTransportProcedures converts encoding-level procedures to transport-level procedures
+func MapToTransportProcedures(encodingProcedures []yarpc.EncodingProcedure) []yarpc.TransportProcedure {
+	transportProcedures := []yarpc.TransportProcedure{}
+	for _, p := range encodingProcedures {
+		var transportHandlerSpec yarpc.TransportHandlerSpec
+		switch p.HandlerSpec.Type() {
+		case yarpc.Unary:
+			transportHandlerSpec = yarpc.NewUnaryTransportHandlerSpec(unaryTransportHandler{p})
+		default:
+			panic(p.HandlerSpec.Type())
+		}
+
+		transportProcedures = append(
+			transportProcedures,
+			yarpc.TransportProcedure{
+				Name:        p.Name,
+				Service:     p.Service,
+				HandlerSpec: transportHandlerSpec,
+				Encoding:    p.Encoding,
+				Signature:   p.Signature,
+			},
+		)
+	}
+
+	return transportProcedures
+}
+
+// NewMapRouterWithProcedures constructs a new MapRouter with the given default service name and registers
+// the given transport-level procedures
+func NewMapRouterWithProcedures(defaultService string, transportProcedures []yarpc.TransportProcedure) MapRouter {
+	router := NewMapRouter(defaultService)
+	router.Register(transportProcedures)
+	return router
+}
+
+// Allows encoding-level procedures to act as transport-level procedures
+type unaryTransportHandler struct {
+	h yarpc.EncodingProcedure
+}
+
+func (u unaryTransportHandler) Handle(ctx context.Context, req *yarpc.Request, reqBuf *yarpc.Buffer) (*yarpc.Response, *yarpc.Buffer, error) {
+	decodedBody, codecErr := u.h.Codec.Decode(reqBuf)
+	if codecErr != nil {
+		return nil, nil, codecErr
+	}
+
+	body, err := u.h.HandlerSpec.Unary().Handle(ctx, decodedBody)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	encodedBody, codecErr := u.h.Codec.Encode(body)
+	if codecErr != nil {
+		return nil, nil, codecErr
+	}
+
+	return nil, encodedBody, nil
```

Comment:
I think we want to return a non-nil `*yarpc.Response` here.

Reply:
I think so too. I believe we can extract it from the Context. Would you happen to know the key for the request meta?

Reply:
Not the request meta, we need a `Response`.

The encoding layer is responsible for creating a `yarpc.Response` and corresponding `yarpc.Buffer` and passing that up to the transport layer.

I _believe_ that this should come from your `InboundCodec` (aka whatever encoding implementation we're using) since each encoding handles this differently.

Reply:
This is still a gap @vyshah. The transport layer will expect a `yarpc.Response` from this layer to send it back to the client.